docs/proposals/20250513-chained-and-efficient-upgrades-for-clusters-with-managed-topologies.md

```diff
@@ -0,0 +1,421 @@
+from today.
+
+This is specifically important for "upgrade workers", which by default is upgrading MachineDeployments first, 
+one after the other according to the order defined in `cluster.spec.topology`, and then MachinePools in a similar way.
+
+Notably, also the exiting options to influence "upgrade workers" will continue to work as of today: 
+- `topology.cluster.x-k8s.io/upgrade-concurrency` annotation can be used to configure the maximum concurrency while
+  upgrading MachineDeployments of a classy Cluster. It is set as a top level annotation on the Cluster object.
+- `topology.cluster.x-k8s.io/defer-upgrade` annotation can be used to defer the Kubernetes upgrade of MachineDeployments
+  in `Cluster.spec.topology.workers`.
+- `topology.cluster.x-k8s.io/hold-upgrade-sequence` can be used to hold the upgrade sequence for a MachineDeployment
+  in `Cluster.spec.topology.workers` and all the following ones.
+
+It also remains unchanged the fact that an "upgrade workers" operation cannot complete until all the MachineDeployment
+are upgraded to the intermediate/target version of this iteration, which is possible only after all the `topology.cluster.x-k8s.io/defer-upgrade` and 
+`topology.cluster.x-k8s.io/hold-upgrade-sequence` annotations are removed.
+
+However, it might be worth to notice that:
+- While performing different "upgrade workers" operations, the target version all MachineDeployments should upgrade to also changes.
+- `topology.cluster.x-k8s.io/defer-upgrade` and `topology.cluster.x-k8s.io/hold-upgrade-sequence` annotations,  must be
+  applied before each upgrade operation (lifecycle hooks described in the next paragraph can be used to orchestrate this process).
+
+#### Lifecycle hooks
+
+As of today, Cluster API implements three lifecycle hooks for upgrades:
+- `BeforeClusterUpgrade`
+- `AfterControlPlaneUpgrade`
+- `AfterClusterUpgrade`
+
+By introducing chained upgrades, it is necessary to expand this set to allow proper tracking of the intermediate steps.
+
+![runtime-hooks](images/lifecycle-hooks/lifecycle-hooks-upgrade.png)
+
+More specifically:
+- `BeforeClusterUpgrade` request will include the upgrade plan.
+- A new `BeforeControlPlaneUpgrade` hook will be added and called before each "upgrade control plane" step;
+  request and response payload will be similar to corresponding messages for `BeforeClusterUpgrade`
+- `AfterControlPlaneUpgrade` hook will now be called after each "upgrade control plane" step.
+- A new `BeforeWorkersUpgrade` hook will be added and called before each "upgrade workers" step;
+  request and response payload will be similar to corresponding messages for `BeforeControlPlaneUpgrade`
+- A new `AfterWorkersUpgrade` hook will be added and called after each "upgrade workers" step;
+  request and response payload will be similar to corresponding messages for `AfterControlPlaneUpgrade`, but the
+  hook will be considered blocking only for the intermediate steps of the upgrade. 
```

Comment:
This sounds like it is not blocking something else. Not sure if I got that right